core/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java

```diff
@@ -399,21 +420,41 @@ public void onFailure(Throwable e) {
                 logger.warn("Failed to update shard information for ClusterInfoUpdateJob within 15s timeout");
             }
 
+            indexClassification = IndexClassification.classifyIndices(indexSizes, logger);
+
+            final ClusterInfo newClusterInfo = getClusterInfo();
             for (Listener l : listeners) {
                 try {
-                    l.onNewInfo(getClusterInfo());
+                    l.onNewInfo(newClusterInfo);
                 } catch (Exception e) {
                     logger.info("Failed executing ClusterInfoService listener", e);
                 }
             }
-        }
-    }
 
-    /**
-     * Method that incorporates the ShardId for the shard into a string that
-     * includes a 'p' or 'r' depending on whether the shard is a primary.
-     */
-    public static String shardIdentifierFromRouting(ShardRouting shardRouting) {
-        return shardRouting.shardId().toString() + "[" + (shardRouting.primary() ? "p" : "r") + "]";
+            clusterService.submitStateUpdateTask("update_cluster_info", new ClusterStateUpdateTask() {
```

Comment:
I am not entirely sure how this information is going to be used on non-master nodes, so this might not be a problem. But since this task has `NORMAL` priority it will tend to get stuck behind `URGENT` and `IMMEDIATE` priority routing-related update tasks during any massive shard-related event (cluster startup, rolling restart, etc.) or might not reach any other nodes at all if the current master will die during such an event. On a relevant note, since these tasks will be accumulating at the end of the pending tasks queue during such events, it might make sense to add some sort of batching so we would only update the cluster with the latest information instead of processing information that was stuck in the queue for a while.